### PR TITLE
Python: Make SK compatible with OpenAI 1.0

### PR DESCRIPTION
**Pull request by alexchaomander**
Need to update SK Python to integrate the breaking changes from openai-python 1.0

https://github.com/microsoft/semantic-kernel/issues/3330

This PR replaces the old one that was merging to main https://github.com/microsoft/semantic-kernel/pull/3417

---------

### Contribution Checklist

- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

### CODE REVIEW

File: python/semantic_kernel/__init__.py [hunk 1]
```diff
@@ -15,7 +15,6 @@
 )
 from semantic_kernel.utils.null_logger import NullLogger
 from semantic_kernel.utils.settings import (
-    azure_cosmos_db_settings_from_dot_env,
```
Comment (moonbox3) [inline python/semantic_kernel/__init__.py:18]: @alexchaomander was this an intentional removal?

File: python/semantic_kernel/connectors/ai/hugging_face/services/hf_text_completion.py [hunk 2]
```diff
@@ -109,28 +95,25 @@ async def complete_async(
...
-                    return completions
+                return completions
 
-            else:
-                raise AIException(
-                    AIException.ErrorCodes.InvalidConfiguration,
-                    "Unsupported hugging face pipeline task: only \
-                        text-generation, text2text-generation, and summarization are supported.",
-                )
+            raise AIException(
```
Comment (moonbox3) [inline python/semantic_kernel/connectors/ai/hugging_face/services/hf_text_completion.py:112]: nit: wouldn't it be better to do this check before we try to do any work? If the task is not defined properly, why spend resources on anything? Perhaps we don't need to fix now, but we could in a next iteration.

File: python/semantic_kernel/connectors/ai/open_ai/services/azure_chat_completion.py [hunk 3]
```diff
@@ -2,67 +2,162 @@
...
+class AzureChatCompletion(
+    AzureOpenAIConfigBase, OpenAIChatCompletionBase, OpenAITextCompletionBase
+):
+    """Azure Chat completion class."""
+
+    @overload
+    def __init__(
+        self,
+        base_url: str,
+        api_version: str = "2022-12-01",
```
Comment (moonbox3) [inline python/semantic_kernel/connectors/ai/open_ai/services/azure_chat_completion.py:32]: Doc string says default value is: 2023-03-15-preview. Should we update?

File: python/semantic_kernel/connectors/ai/open_ai/services/azure_chat_completion.py [hunk 4]
```diff
@@ -2,67 +2,162 @@
...
+            logger: deprecated, use 'log' instead.
+        """
 
+    @overload
     def __init__(
         self,
         deployment_name: str,
-        endpoint: Optional[str] = None,
+        endpoint: str,
+        api_version: str = "2022-12-01",
```
Comment (moonbox3) [inline python/semantic_kernel/connectors/ai/open_ai/services/azure_chat_completion.py:66]: Similar question as above: docstring shows a different api version?

File: python/semantic_kernel/connectors/ai/open_ai/services/azure_config_base.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+from semantic_kernel.connectors.ai.open_ai.services.open_ai_handler import (
+    OpenAIHandler,
+    OpenAIModelTypes,
+)
+from semantic_kernel.sk_pydantic import HttpsUrl
+
+
+class AzureOpenAIConfigBase(OpenAIHandler):
+    @validate_arguments(config=dict(arbitrary_types_allowed=True))
+    def __init__(
```
Comment (moonbox3) [inline python/semantic_kernel/connectors/ai/open_ai/services/azure_config_base.py:19]: nit: should we include a docstring? it seems like it's useful in this context.

File: python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        response = await self._send_request(
+            messages=messages, request_settings=settings, stream=False
+        )
+
+        if len(response.choices) == 1:
+            return response.choices[0].message.content
+        else:
+            return [choice.message.content for choice in response.choices]
+
+    async def complete_chat_with_functions_async(
```
Comment (moonbox3) [inline python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py:52]: nit: nit missing docstring?

File: python/semantic_kernel/utils/settings.py [hunk 7]
```diff
@@ -198,24 +198,6 @@ def google_palm_settings_from_dot_env() -> str:
     return api_key
 
 
-def azure_cosmos_db_settings_from_dot_env() -> Tuple[str, str]:
```
Comment (moonbox3) [inline python/semantic_kernel/utils/settings.py:201]: We should make sure this isn't removed.

File: python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        Returns:
+            Union[str, List[str]] -- The completion result(s).
+        """
+        response = await self._send_request(
+            messages=messages, request_settings=settings, stream=True
+        )
+
+        # parse the completion text(s) and yield them
+        async for chunk in response:
+            text, index = self._parse_choices(chunk)
```
Comment (moonbox3) [inline python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py:100]: Based on the following PR, we'd need to make some updated to handle parsing choices for streaming respsonses: https://github.com/microsoft/semantic-kernel/pull/3651.

**Review by moonbox3 — APPROVED** [review]

**Review by markwallace-microsoft — APPROVED** [review]

**Review by moonbox3 — APPROVED** [review]
Fixes to integration tests just went in. Approving.

File: python/semantic_kernel/connectors/ai/google_palm/services/gp_chat_completion.py [hunk 9]
```diff
@@ -64,11 +62,9 @@ async def complete_chat_async(
                 else "I don't know."
                 for candidate in response.candidates
             ]
-        else:
-            if response.last is None:
-                return "I don't know."  # PaLM returns None if it doesn't know
-            else:
-                return response.last
+        if response.last is None:
+            return "I don't know."  # PaLM returns None if it doesn't know
```
Comment (TaoChenOSU) [inline python/semantic_kernel/connectors/ai/google_palm/services/gp_chat_completion.py:66]: nit: not from this change but wouldn't it be better if this was a constant?

**Review by TaoChenOSU — APPROVED** [review]

File: python/semantic_kernel/connectors/ai/google_palm/services/gp_chat_completion.py (hunk 9 above)
Comment (moonbox3) [reply python/semantic_kernel/connectors/ai/google_palm/services/gp_chat_completion.py:66]: Thanks, we can address in a next iteration.

File: python/semantic_kernel/connectors/ai/open_ai/const.py [hunk 10]
```diff
@@ -0,0 +1,3 @@
+from typing import final
```
Comment (dluc) [inline python/semantic_kernel/connectors/ai/open_ai/const.py:1]:
```suggestion
# Copyright (c) Microsoft. All rights reserved.

from typing import final
```

File: python/semantic_kernel/connectors/ai/open_ai/services/open_ai_model_types.py [hunk 11]
```diff
@@ -0,0 +1,9 @@
+from enum import Enum
```
Comment (dluc) [inline python/semantic_kernel/connectors/ai/open_ai/services/open_ai_model_types.py:1]:
```suggestion
# Copyright (c) Microsoft. All rights reserved.

from enum import Enum
```

**Review by dluc — CHANGES_REQUESTED** [review]
a couple of missing headers

**Review by dluc — APPROVED** [review]

**Review by markwallace-microsoft — APPROVED** [review]

**Pull request merged**